### PR TITLE
CDAP-20342 : Spark SQL engine implementation as a fallback engine

### PR DESCRIPTION
**Pull request by sahusanket**

### CODE REVIEW

File: cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/io/cdap/cdap/etl/spark/batch/BatchSparkPipelineDriver.java [hunk 1]
```diff
@@ -413,11 +419,12 @@ protected Iterable<SparkCollectionRelationalEngine> getRelationalEngines(StageSp
       return super.getRelationalEngines(stageSpec, stageData);
     }
 
+    //TODO : Decide on this
     // If this stage is not pushed down and it's not in the included stages, we can skip relational transformation
     // in the SQL engine.
-    if (!(stageData instanceof SQLBackedCollection) && !shouldForcePushToSQLEngine(stageSpec.getName())) {
-      return super.getRelationalEngines(stageSpec, stageData);
-    }
+//    if (!(stageData instanceof SQLBackedCollection) && !shouldForcePushToSQLEngine(stageSpec.getName())) {
```
Comment (fernst) [inline cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/io/cdap/cdap/etl/spark/batch/BatchSparkPipelineDriver.java]: I think we need 2 pieces of logic here:

1. Check if the SQLBackedCollection is compatible with the engine. For example, a BigQuery SQLBackedCollection is not compatible with Spark, so this needs to be executed in Spark.
2. If the stage is not pushed down, we should push down, but only when the "cost" of the push operation is zero.

I see a couple things to consider here

1. We need to ensure that the `SQLBackedCollection` is compatible with the engines we return. In general, if the collection is either a Spark of a BigQuery collection, the only engine that can handle it is the corresponding engine.


2. If the engine is Spark, the cost to push/pull is pretty much zero. We can update the SQL Engine Adapter with a new configuration property to decide if the SQL Engine is local or remote. If the engine is initialized by the customer using the Pushdown configuration, we can assume the engine is running remotely. If the engine is initialized as a fallback (Spark), we can assume the engine is local, and the cost of push/pull is minimal and should not be considered relevant.

```
// Add to BatchSQLEngineAdapter class 
boolean isLocalEngine() {
  return true; // if spark
  return false; // if any other engine
}
```

If this property is true, and the state is not pushed before the relational transform, we can "push" (for little/no cost) and perform the relational transform operation.

File: cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/io/cdap/cdap/etl/spark/batch/SQLEngineRelationalEngine.java [hunk 2]
```diff
@@ -67,8 +67,16 @@ public Engine getRelationalEngine() {
...
-    return adapter.tryRelationalTransform(stageSpec, transform, input).map(job ->
-      new SQLEngineCollection<T>(sec, functionCacheFactory, jsc, sqlContext,
-                                 datasetContext, sinkFactory, stageSpec.getName(), adapter, job));
+     return adapter.tryRelationalTransform(stageSpec, transform, input)
+      .map(job -> {
+        SQLEngineCollection<T> sqlEngineCollection = new SQLEngineCollection<T>(sec, functionCacheFactory, jsc,
+                                                                                sqlContext, datasetContext, sinkFactory,
+                                                                                stageSpec.getName(), adapter, job);
+        // If it is a Spark Sql Job with a local engine, then Perform a Pull resulting in RddCollection.
+        if (adapter.getSQLEngineClassName().equals(SparkSQLEngine.class.getName())) {
```
Comment (fernst) [inline cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/io/cdap/cdap/etl/spark/batch/SQLEngineRelationalEngine.java:76]: How does this work if Pushdown is enabled? As far as I can tell, the adapter is not aware of the fallback engine.

Comment (sahusanket) [reply cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/io/cdap/cdap/etl/spark/batch/SQLEngineRelationalEngine.java:76]: It's the adaptor that is being used as a fall back tool. 
In `BatchSparkPipelineDriver` there's `sqlEngineAdapter` and a `fallbackSqlEngineAdapter` .

Comment (fernst) [reply cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/io/cdap/cdap/etl/spark/batch/SQLEngineRelationalEngine.java:76]: Ah, I see.

**Review by fernst — APPROVED** [review]

**Pull request merged**